Firestore/core/src/firebase/firestore/util/to_string.h

```diff
@@ -0,0 +1,218 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_TO_STRING_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_TO_STRING_H_
+
+#if __OBJC__
+#import <Foundation/Foundation.h>
+#endif  // __OBJC__
+
+#include <algorithm>
+#include <numeric>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+#include "Firestore/core/src/firebase/firestore/util/type_traits.h"
+#include "absl/meta/type_traits.h"
+#include "absl/strings/str_join.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+template <typename T>
+std::string ToString(const T& value);
+
+namespace impl {
+
+// Checks whether the given type `T` defines a member function `ToString`
+
+template <typename T, typename = absl::void_t<>>
+struct has_to_string : std::false_type {};
+
+template <typename T>
+struct has_to_string<T, absl::void_t<decltype(std::declval<T>().ToString())>>
+    : std::true_type {};
+
+// Fallback
+
+template <typename T>
+std::string DefaultToString(const T& value) {
+  FormatArg arg{value};
+  return std::string{arg.data(), arg.data() + arg.size()};
+}
+
+// Container
+
+template <typename T>
+std::string ContainerToString(const T& value, std::false_type) {
+  return DefaultToString(value);
+}
+
+template <typename T>
+std::string ContainerToString(const T& value, std::true_type) {
+  std::string contents = absl::StrJoin(
+      value, ", ", [](std::string* out, const typename T::value_type& element) {
+        out->append(ToString(element));
+      });
+  return std::string{"["} + contents + "]";  // NOLINT(whitespace/braces)
+}
+
+// Associative container
+
+template <typename T>
+std::string MapToString(const T& value, std::false_type) {
+  return ContainerToString(value, is_iterable<T>{});
+}
+
+template <typename T>
+std::string MapToString(const T& value, std::true_type) {
+  std::string contents = absl::StrJoin(
+      value, ", ", [](std::string* out, const typename T::value_type& kv) {
+        out->append(
+            StringFormat("%s: %s", ToString(kv.first), ToString(kv.second)));
+      });
+  return std::string{"{"} + contents + "}";  // NOLINT(whitespace/braces)
+}
+
+// std::string
+
+template <typename T>
+std::string StringToString(const T& value, std::false_type) {
+  return MapToString(value, is_associative_container<T>{});
+}
+
+template <typename T>
+std::string StringToString(const T& value, std::true_type) {
+  return value;
+}
+
+#if __OBJC__
+
+// Objective-C class
+
+template <typename T>
+std::string ObjCToString(const T& value, std::false_type) {
```

Comment:
> Do you think it would be clearer to always define `ObjcToString` but if not `__OBJC__` always delegate down?

Another way would be something like
```cpp
template <typename T>
std::string ObjCToString(const T& value, std::false_type) {
  return StringToString(value, std::is_same<T, std::string>{});
}

#if __OBJC__
template <typename T>
std::string ObjCToString(const T& value, std::true_type) {
  return MakeString([value description]);
}
#endif

template <typename T>
std::string CustomToString(const T& value, std::false_type) {
  // This would require a different is_objective_c_pointer definition
  // in pure C++, or another ifdef block here
  return ObjCToString(value, is_objective_c_pointer<T>::value);
}
```


I don't have a strong preference on this mechanism versus `FormatChoice`. I agree with all your points except the last one (to me, there's no significant difference between `FormatChoice<0>` and `std::true_type`). It's certainly good for readability that the order isn't reversed and there's less boilerplate. On the other hand, the way `FormatChoice` works is more complicated. There's recursive inheritance, but more significantly, every function there is a potential candidate for the overload set, so it's necessary to remember to apply `enable_if` appropriately (but not always necessary, e.g., the `const char*` overload doesn't have it). Also, adding a new element to the chain would require modifying several indices as well as their total number (though it's not a huge deal since the number is small).

I tried out the `FormatChoice` approach, what do you think? Since I don't have a strong preference, I'll go with whichever you prefer.